src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java

```diff
@@ -4,35 +4,221 @@
+		if(rightSpeed >= RobotMap.VELOCITY_LOWER_SOFT_LIMIT && rightSpeed <= RobotMap.VELOCITY_UPPER_SOFT_LIMIT)
+		{
+			rightFrontTalon.changeControlMode(TalonControlMode.Speed);
+		
+			rightFrontTalon.set(rightSpeed);
+		}
+	}
+	
+	public void driveRightPercent(double rightSpeed) {
+		if(rightSpeed >= RobotMap.PERCENT_LOWER_SOFT_LIMIT && rightSpeed <= RobotMap.PERCENT_UPPER_SOFT_LIMIT )
+		{
+			rightFrontTalon.changeControlMode(TalonControlMode.PercentVbus);
+		
+			rightFrontTalon.set(rightSpeed);
+		}
 	}
 
-	public void driveRight(double rightSpeed) {
-		rightFrontTalon.set(rightSpeed * RobotMap.DRIVETRAIN_FORWARD_RIGHT_CONSTANT);
-		rightBackTalon.set(rightSpeed * RobotMap.DRIVETRAIN_FORWARD_RIGHT_CONSTANT);
+	
+	
+	/**
+	 * Gets the current position of the robot
+	 * Multiplied by the circumference of the wheel for scaling
+	 * @return
+	 */
+	public double getLeftPosition(){
+		return leftFrontTalon.getPosition() * RobotMap.DRIVETRAIN_WHEEL_CIRCUMFERENCE_CONSTANT;
+	}
+	
+	public double getRightPosition(){
+		return rightFrontTalon.getPosition() * RobotMap.DRIVETRAIN_WHEEL_CIRCUMFERENCE_CONSTANT;
+	}
+	
+	/**
+	 * Methods that get/set the left and right positions of where we want to be
+	 * setPosition zeros out the encoder data, so that we start at 0
+	 */
+	
+	public void setLeftDesiredPosition(double goalDistance){
+		leftFrontTalon.setPosition(0);
+		this.goalDistance = goalDistance;
+	}
+	
+	public void setRightDesiredPosition(double goalDistance){
+		rightFrontTalon.setPosition(0);
+		this.goalDistance = goalDistance;
+	}
+	
+	public double getLeftDesiredDistance()
+	{
+		return goalDistance;
+	}
+	
+	public double getRightDesiredDistance()
+	{
+		return goalDistance;
 	}
+	
 
+	
+	/**
+	 * Methods that calculate the error in distance from where we are to where we want to be
+	 */
+	
+	public double calculateLeftDistanceError(){
+		return rightFrontTalon.getError();
+		
+	}
+	
+	public double calculateRightDistanceError(){
+		return leftFrontTalon.getError();
+		
+	}
+	
+	/**
+	 * Methods to increment, decrement, and get the distance constant
+	 * Method takes in the talon the P constant is being set for, gets the current P, and sets
+	 * the P to the current P + the increment/decrement constant
+	 */
+	
+	public void incrementDistanceConstant(CANTalon currentTalon){
+		double currentPConstant = currentTalon.getP();
+		currentTalon.setP(currentPConstant += RobotMap.DISTANCE_INCREMENT_CONSTANT);
+	}
+	
+	public void decrementDistanceConstant(CANTalon currentTalon){
+		double currentPConstant = currentTalon.getP();
+		currentTalon.setP(currentPConstant-= RobotMap.DISTANCE_DECREMENENT_CONSTANT);
+	}
+	
+	public double returnDistanceConstant(CANTalon currentTalon){
+		return currentTalon.getP();
+	}
+	
+	/**
+	 * Sets desired angle for driving at an angle
+	 * Currently unfinished
+	 * @param goalAngle
+	 */
+	public void setDesiredAngle(double goalAngle) {
+		desiredAngle = goalAngle;
+	}
+	
+	public void moveToAngle() {
```

Comment:
Two issues with this one:
1. The method is currently a stub, so I don't see why it should exist at all if it's empty.
2. If this is intended for what I believe it is— rotating the robot in the xy plane— it should not be in the Subsystem. A `Subsystem` is a programmatic representation of the physical robot subsystem, and is meant only to provide methods for directly controlling that subsystem: For example, "make these motors spin at this RPM". In other words, it establishes a programmatic view into **what is possible with the subsystem**. This method is better suited for a `Command`: It represents a task that we want the robot to do, and instructs the `DriveTrain` on how to do it.